keyvalue/clients/src/main/java/tachyon/client/keyvalue/hadoop/KeyValueRecordReader.java

```diff
@@ -0,0 +1,123 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.client.keyvalue.hadoop;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.mapred.RecordReader;
+
+import tachyon.client.keyvalue.KeyValueIterator;
+import tachyon.client.keyvalue.KeyValuePair;
+import tachyon.client.keyvalue.KeyValuePartitionReader;
+import tachyon.exception.TachyonException;
+import tachyon.util.io.BufferUtils;
+
+/**
+ * Implements {@link RecordReader}, each record is a key-value pair stored in a partition of the
+ * {@link tachyon.client.keyvalue.KeyValueStores}.
+ */
+final class KeyValueRecordReader implements RecordReader<BytesWritable, BytesWritable> {
+  /** The partition reader for reading the key-value pairs */
+  private final KeyValuePartitionReader mReader;
+  /** The iterator for iterating through all key-value pairs contained in the partition */
+  private final KeyValueIterator mKeyValuePairIterator;
+  /** Current position in the partition's byte array */
+  private int mPos;
+  /** Number of key-value pairs visited by the iterator */
+  private int mNumVisitedKeyValuePairs;
+  /** Number of key-value pairs */
+  private final int mNumKeyValuePairs;
+
+  /**
+   * Creates a {@link KeyValueRecordReader} for generating key-value pairs of a partition.
+   *
+   * @param split the split for a block
+   * @throws IOException if non-Tachyon error occurs
+   * @throws TachyonException if Tachyon error occurs
+   */
+  public KeyValueRecordReader(KeyValueInputSplit split) throws IOException, TachyonException {
+    mReader = KeyValuePartitionReader.Factory.create(split.getPartitionId());
+    mKeyValuePairIterator = mReader.iterator();
+    mPos = 0;
+    mNumVisitedKeyValuePairs = 0;
+    mNumKeyValuePairs = mReader.size();
+  }
+
+  @Override
+  public boolean next(BytesWritable keyWritable, BytesWritable valueWritable) throws IOException {
+    if (!mKeyValuePairIterator.hasNext()) {
+      return false;
+    }
+
+    KeyValuePair pair;
+    try {
+      pair = mKeyValuePairIterator.next();
+    } catch (TachyonException e) {
+      throw new IOException(e);
+    }
+
+    DataInputStream key = new DataInputStream(new ByteArrayInputStream(
+        BufferUtils.newByteArrayFromByteBuffer(pair.getKey())));
+    try {
+      keyWritable.readFields(key);
+    } finally {
+      key.close();
+    }
+
+    DataInputStream value = new DataInputStream(new ByteArrayInputStream(
+        BufferUtils.newByteArrayFromByteBuffer(pair.getValue())));
+    try {
+      valueWritable.readFields(value);
+    } finally {
+      value.close();
+    }
+
+    mPos += keyWritable.getLength() + valueWritable.getLength();
+    mNumVisitedKeyValuePairs ++;
+    return true;
+  }
+
+  @Override
+  public BytesWritable createKey() {
+    return new BytesWritable(new byte[0]);
```

Comment:
It's not necessary, just `new BytesWritable()` is enough, I've updated the code.